light-client/src/storage.rs

```diff
@@ -0,0 +1,334 @@
+
+        Ok(SqliteStorage {
+            pool,
+            num_leaves: self.num_leaves,
+        })
+    }
+}
+
+/// [`Storage`] based on a SQLite database.
+#[derive(Clone, Debug)]
+pub struct SqliteStorage {
+    pool: SqlitePool,
+    num_leaves: u32,
+}
+
+impl Storage for SqliteStorage {
+    async fn default() -> Result<Self> {
+        SqliteOptions::default().connect().await
+    }
+
+    async fn leaf_upper_bound(
+        &self,
+        id: LeafId<SeqTypes>,
+    ) -> Result<Option<LeafQueryData<SeqTypes>>> {
+        let mut tx = self.pool.begin().await?;
+        let (height, data): (i64, _) = match id {
+            LeafId::Number(n) => {
+                let Some((height, data)) = query_as(
+                    "SELECT height, data FROM leaf WHERE height >= $1 ORDER BY height LIMIT 1",
+                )
+                .bind(n as i64)
+                .fetch_optional(tx.as_mut())
+                .await?
+                else {
+                    return Ok(None);
+                };
+                (height, data)
+            },
+            LeafId::Hash(h) => {
+                let Some((height, data)) =
+                    query_as("SELECT height, data FROM leaf WHERE hash = $1 LIMIT 1")
+                        .bind(h.to_string())
+                        .fetch_optional(tx.as_mut())
+                        .await?
+                else {
+                    return Ok(None);
+                };
+                (height, data)
+            },
+        };
+        let leaf = serde_json::from_value(data)?;
+
+        // Mark this leaf as recently used.
+        let (id,): (i32,) = query_as("SELECT max(id) + 1 FROM leaf")
+            .fetch_one(tx.as_mut())
+            .await?;
+        query("UPDATE leaf SET id = $1 WHERE height = $2")
+            .bind(id)
+            .bind(height)
+            .execute(tx.as_mut())
+            .await?;
+        tx.commit().await?;
+
+        Ok(Some(leaf))
+    }
+
+    async fn insert_leaf(&self, leaf: LeafQueryData<SeqTypes>) -> Result<()> {
+        let mut tx = self.pool.begin().await?;
+
+        let height = leaf.height() as i64;
+        let hash = leaf.hash().to_string();
+        let data = serde_json::to_value(leaf)?;
+
+        tracing::debug!(height, hash, "inserting leaf");
+        let (id,): (i32,) = query_as(
+            "INSERT INTO leaf (height, hash, data) VALUES ($1, $2, $3)
+                    ON CONFLICT (height) DO UPDATE SET id = excluded.id
+                    RETURNING id",
+        )
+        .bind(height)
+        .bind(&hash)
+        .bind(data)
+        .fetch_one(tx.as_mut())
+        .await
+        .context("inserting new leaf")?;
+        tracing::debug!(height, hash, id, "inserted leaf");
+
+        // Delete the oldest leaves as necessary until the number of leaves stored does not exceed
+        // `num_leaves`.
+        let (num_leaves,): (u32,) = query_as("SELECT count(*) FROM leaf")
```

Comment:
nit: Doing garbage collection while inserting leaf might slow down insertions, maybe we could do garbage collection in the background?

Reply:
I think it is just a couple extra queries against a database that is on the local file system or even in memory. It doesn't involve any network round trips or anything. I'm hoping this will be fine.

I would wait to see that it's actually a problem before adding the complexity of async garbage collection